### PR TITLE
Logger: added tool for parsing new trace format

### PR DESCRIPTION
**Pull request by bkokoszx**
I've added tool for parsing new trace format.
As input tool gets two files:
1) extracted .static_log_entries section generated by rimage in file *ldc (sof\src\arch\xtensa\*ldc) - this rimage feature is added in #418 PR in sof repository;
2) binary file containing catch trace from dma.
Tool throws parsed logs to stdout.

### CODE REVIEW

File: logger/logger.c [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+struct ldc_entry {
+	struct ldc_entry_header header;
+	char *file_name;
+	uint32_t text_len;
+	char *text;
+	uint32_t *params;
+}; 
+
+struct dma_log_header {
```
[the remark this answers is not in the document]
Comment (bkokoszx) [reply logger/logger.c]: @lgirdwood Ok, but in what way can I include uapi headers existing in different repo? (sof)

[the remark this answers is not in the document]
Comment (lgirdwood) [reply logger/logger.c]: @bkokoszx the should be installed by "make install" rule from FW.

File: logger/logger.c [hunk 2]
```diff
@@ -0,0 +1,275 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <errno.h>
+
+#define CEIL(a, b) ((a+b-1)/b)
```
[the remark this answers is not in the document]
Comment (bkokoszx) [reply logger/logger.c]: Yes, in line:
/* padding - sequences of chars are aligned to DWORDS */
	fseek(f_ldc, CEIL(entry.header.file_name_len, sizeof(uint32_t)) *
		sizeof(uint32_t) - entry.header.file_name_len, SEEK_CUR);

File: logger/logger.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (bkokoszx) [reply logger/logger.c]: I've included from "/usr/local/include/sof/uapi/logging.h"

File: logger/logger.c [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+	entry_offset = dma_log.address - base_address;
+	
+	/* set file position to beginning of processed entry */
+	fseek(f_ldc, entry_offset + data_offset, SEEK_SET);
+	
+	/* fetching elf header params */
+	ret = fread(&entry.header, sizeof(entry.header), 1, f_ldc);
+
+	if (!ret) {
+		ret = -EINVAL;
```
[the remark this answers is not in the document]
Comment (bkokoszx) [reply logger/logger.c]: I've added returning errno to handle fopen function. I've read that fread doesn't set errno

**Comment by bkokoszx** [comment]
travis-ci was expected to fail. Sof should be installed ("make install") before building soft.

**Comment by xiulipan** [comment]
@bkokoszx 
Why we need sof install for soft?
We want SOFT to be install independently. Thus why we move rimage into sof to avoid this situation.

**Comment by bkokoszx** [comment]
@xiulipan 
I have to exploit struct existing in uapi header in FW. I can use use it since I install FW (make install copy headers to: /usr/local/include/sof/uapi/). It was change due Liam suggestion (https://github.com/thesofproject/soft/pull/77#discussion_r220581157).
@lgirdwood 
Did I misunderstand something?

**Comment by xiulipan** [comment]
@lgirdwood 
So now we need some dependency on SOF again?
What should this work with trace event from @bardliao  then? And the trace level from @RanderWang

**Comment by singalsu** [comment]
@bkokoszx @lgirdwood Is it possible for you to remove the offending code from SOF if this work is not mature enough to be merged to kernel and SOFT.

File: logger/logger.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply logger/logger.c]: usual convention is to use sizeof and % operator to make it obvious we are aligning to a size

**Comment by lgirdwood** [comment]
@xiulipan soft tools can depend on sof. rimage was moved as dependency was the other way around. The tools need the uapi headers to use the ABI.

@bkokoszx can you resend new PR today and ignore the ASCII HEX readback part, this wont be ready for at least a week since national holidays in China.

**Comment by bkokoszx** [comment]
@lgirdwood 
I have return errno only to handle fopen functions. I've read that fread function doesn't set errno, so in that case I return EINVAL. Am I wrong?

**Comment by xiulipan** [comment]
@lgirdwood 
So how do we handler the CI issues?
If we need to git clone SOF first?

@bkokoszx 
And how about the old rmbox? should this logger to replace the rmbox or rmbox will still alive?
Maybe you can just rewrite the rmbox with the rmbox help functions.

File: logger/logger.c [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+int main(int argc, char *argv[])
+{
+	struct dma_log dma_log;
+	struct snd_sof_logs_header snd;
+
+	int ret;
+
+	const char *ldc_dir = argv[1];
+	const char *dma_dump = argv[2];
```
Comment (xiulipan) [inline logger/logger.c]: Can you take rmbox as reference to have use getopt for the function.
Or just replace the old rmbox.

```c
	while ((opt = getopt(argc, argv, "ho:i:s:m:c:t")) != -1) {
		switch (opt) {
		case 'o':
			out_file = optarg;
			break;
		case 'i':
			in_file = optarg;
			break;
		case 't':
			trace = 1;
			break;
		case 'c':
			clk = atof(optarg);
			break;
		case 's':
			return snapshot(optarg);
		case 'h':
		default: /* '?' */
			usage(argv[0]);
		}
	}
```

Comment (lgirdwood) [reply logger/logger.c]: @bkokoszx yes pls use getopt, no need to replace rmbox atm as we need it for legacy. I want to merge this PR today.

**Comment by bkokoszx** [comment]
To handle fread I use following code:
ret = fread(..., file);
	if (!ret) {
		ret = -ferror(file);
		goto out;
	}
I will update getops

**Comment by bkokoszx** [comment]
@lgirdwood 
I've updated PR using getops function

**Comment by lgirdwood** [comment]
@bkokoszx one more change that can be done incrementally is to use relative path for uapi/logger.h i.e. include <sof/uapi/logger.h>

**Comment by cujomalainey** [comment]
Synchronization through version matching in the ABI?

**Comment by lgirdwood** [comment]
Yes, we would need some version data. @bkokoszx can you add versioning support here and we can duplicate the header.

**Comment by dabekjakub** [comment]
@mwierzbix @lgirdwood @bkokoszx 
In my experience  having duplicate headers doesn't work very good. Tool repository headers will get updated only when the tool stops working and most of the users won't even be aware of the dependency. Which in worst case can lead to incorrect translation of the logs with no other symptoms.

Another solution to this could be moving logger itself to the FW repository (As was done with rimage). Since then we will have a complete package after building (FW and application to translate logs).

**Comment by lgirdwood** [comment]
@mmaka1 any objections to merging soft -> sof repo ? This would make CI easier too for working topologies since they would be aligned with FW. @cujomalainey would merging repos be OK for you ?

**Comment by cujomalainey** [comment]
Would it be just the tools migrated or the topologies as well? Either should be fine, I don't see any reason why that would be an issue over here. We will just need to update the scripts and docker system once the change is done.

**Comment by mmaka1** [comment]
@lgirdwood I am Ok with this simplification.

**Comment by bkokoszx** [comment]
@lgirdwood What is your opinion about using git-submodule mentioned by @mwierzbix?

**Comment by mwierzbix** [comment]
@bkokoszx @lgirdwood Some pretty solid articles in the subject:
https://medium.com/@porteneuve/mastering-git-submodules-34c65e940407
https://medium.com/@porteneuve/mastering-git-subtrees-943d29a798ec

**Comment by cujomalainey** [comment]
Just make sure everyone understands to run `git submodule update --init --recursive` otherwise there will be a lot of confusing build failures

**Pull request merged**